From: Nick <nick@foxsec.net>
#Subject: [PATCH 2/2] ntdll: fix LdrGetDllPath with LOAD_WITH_ALTERED_SEARCH_PATH
Message-Id: <9ba30d63-ecee-73c4-f2d9-c5e128cfdb1c@foxsec.net>
Date: Sun, 10 Oct 2021 04:19:44 +0000

Empty Message

From d56161e00c2418ddfa7fb570a06894f570dc39cd Mon Sep 17 00:00:00 2001
From: Nick Fox <nick@foxsec.net>
Date: Sun, 10 Oct 2021 00:11:10 -0400
Subject: [PATCH 2/2] ntdll: fix LdrGetDllPath with
 LOAD_WITH_ALTERED_SEARCH_PATH

Windows adds the executable directory to the path when LdrGetDllPath
is called with LOAD_WITH_ALTERED_SEARCH_PATH and a relative module name,
but only when the module name does not include directories.

Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=26350
Wine-Bug: https://bugs.winehq.org/show_bug.cgi?id=51821

Signed-off-by: Nick Fox <nick@foxsec.net>
---
 dlls/ntdll/loader.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index 797d72aedb6..cda4477c431 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -4208,7 +4208,7 @@ NTSTATUS WINAPI LdrGetDllPath( PCWSTR module, ULONG flags, PWSTR *path, PWSTR *u
     else
     {
         const WCHAR *dlldir = dll_directory.Length ? dll_directory.Buffer : NULL;
-        if (!(flags & LOAD_WITH_ALTERED_SEARCH_PATH))
+        if (!(flags & LOAD_WITH_ALTERED_SEARCH_PATH) || (flags & LOAD_WITH_ALTERED_SEARCH_PATH && !wcschr( module, '\\' )))
             module = NtCurrentTeb()->Peb->ProcessParameters->ImagePathName.Buffer;
         status = get_dll_load_path( module, dlldir, dll_safe_mode, path );
     }

-- 
2.33.0

